File: python/cicada_nn/spread_utils.py
"""
Instrument-specific spread and point size utilities.
Known symbols only. Unknown symbols: log error and raise — use broker pip_size.
"""

import logging
import math
import re

logger = logging.getLogger(__name__)


def infer_point_size(symbol: str, mid_price: float = 1.0) -> float:
    """
    Return point/pip size for known symbol patterns only. No inference for unknowns.
    Known: JPY pairs, synthetics (R_, CRASH, BOOM, etc.), indices (US30, AUS200, etc.).
    Unknown symbols: log error and raise — use broker pip_size.
    """
    sym = symbol.upper().replace(" ", "")
    if "JPY" in sym or re.search(
        r"USDJPY|EURJPY|GBPJPY|AUDJPY|NZDJPY|CADJPY|CHFJPY", sym, re.I
    ):
        return 0.01
    if re.search(r"^R_|^CRASH|^BOOM|^1HZ|^JUMP_|^RANGE_BREAK|^WLD|^STPRNG", sym):
        return 0.01
    if re.search(r"^(US|AU|EU|UK|DE|JP|CH)\d{2,3}$|^[A-Z]{2,4}\d{2,}$", sym):
        return 0.01
    err = f"Unknown symbol for point size: {symbol}. Use broker pip_size; no inference."
    logger.error("[spread_utils] %s", err)
    raise ValueError(err)


def spread_points_to_fraction(
    spread_points: float,
    instrument_symbol: str,
    mid_price: float = 1.0,
) -> float:
    """
    Convert spread from points/pips to price fraction.
    Uses instrument-specific point size — no generic 1 pip = 1e-4 assumption.
    """
    pts = spread_points if isinstance(spread_points, (int, float)) and not math.isnan(spread_points) else 0.0
    point_size = infer_point_size(instrument_symbol, mid_price)
    raw = pts * point_size
    result = min(0.01, max(0.0, raw))
    return result if not math.isnan(result) else 0.0

File: src/app/core/spreadUtils.ts
/**
 * Instrument-specific spread and point size utilities.
 * Avoids simplified one-size-fits-all logic (e.g. 1 pip = 1e-4 for all instruments).
 * JPY pairs use 0.01 per pip; indices/synthetics use 0.01; 4-decimal forex uses 0.0001.
 */

/**
 * Return point/pip size for known symbol patterns only. No inference for unknowns.
 * Known: JPY pairs, synthetics (R_, CRASH, BOOM, etc.), indices (US30, AUS200, etc.).
 * Unknown symbols: log error and throw — use broker pip_size instead.
 */
export function inferPointSize(symbol: string, _midPrice: number = 1): number {
  const sym = symbol.toUpperCase().replace(/\s+/g, '');
  // Deriv API symbols for forex (frx*) / crypto (cry*) from ticks_history
  if (/^FRX/i.test(sym)) {
    if (sym.includes('JPY')) return 0.01;
    return 0.0001;
  }
  if (/^CRY/i.test(sym)) return 0.01;
  // Display forex major pairs (EUR/USD, EURUSD)
  if (/^[A-Z]{3}\/[A-Z]{3}$/.test(sym)) {
    if (sym.includes('JPY')) return 0.01;
    return 0.0001;
  }
  if (sym.includes('JPY') || /USDJPY|EURJPY|GBPJPY|AUDJPY|NZDJPY|CADJPY|CHFJPY/i.test(sym)) return 0.01;
  if (/^R_|^CRASH|^BOOM|^1HZ|^JUMP_|^RANGE_BREAK|^WLD|^STPRNG/i.test(sym)) return 0.01;
  if (/^(US|AU|EU|UK|DE|JP|CH)\d{2,3}$|^[A-Z]{2,4}\d{2,}$/i.test(sym)) return 0.01;
  if (/^(BTC|ETH|SOL|XRP|DOGE)USD$/i.test(sym)) return 0.01;
  // Six-letter majors without slash (EURUSD, GBPJPY) — exclude symbols with digits (indices)
  if (/^[A-Z]{6}$/.test(sym) && !/\d/.test(sym)) {
    if (sym.includes('JPY')) return 0.01;
    return 0.0001;
  }
  const err = `[spreadUtils] Unknown symbol for point size: ${symbol}. Use broker pip_size; no inference.`;
  if (typeof console !== 'undefined' && console.error) console.error(err);
  throw new Error(err);
}

/**
 * Convert spread from points/pips to price fraction.
 * Uses instrument-specific point size — no generic 1 pip = 1e-4 assumption.
 * @param spreadPoints - Spread in points (e.g. 0.8 for forex, 5 for indices)
 * @param instrumentSymbol - Symbol for point size inference (e.g. R_10, EUR/USD)
 * @param midPrice - Optional mid price; when unknown, 1 is used (forex default)
 */
export function spreadPointsToFraction(
  spreadPoints: number,
  instrumentSymbol: string,
  midPrice: number = 1
): number {
  const pts = Number.isFinite(spreadPoints) ? spreadPoints : 0;
  const pointSize = inferPointSize(instrumentSymbol, midPrice);
  const raw = pts * pointSize;
  const result = Math.min(0.01, Math.max(0, raw));
  return Number.isFinite(result) ? result : 0;
}
